modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -89,86 +157,468 @@ public IgniteStatisticsRepository statisticsRepository() {
+    ) throws IgniteCheckedException {
+        final StatsCollectionFutureAdapter doneFut = new StatsCollectionFutureAdapter();
+
+        UUID reqId = UUID.randomUUID();
+        StatsKeyMessage keyMsg = StatisticsUtils.toMessage(schemaName, objName, colNames);
+
+        Map<UUID, List<Integer>> requestNodes = null;
+        /*try {
+            requestNodes = nodePartitions(extractGroups(Collections.singletonList(keyMsg)), null);
+        } catch (IgniteCheckedException e) {
+            // TODO: handle & remove task
+        }
+
+        StatsClearRequest req = new StatsClearRequest(reqId, false, Collections.singletonList(keyMsg));
+
+        sendLocalRequests(reqId, req, requestNodes);*/
+
+        clearObjectStatisticsLocal(keyMsg);
+    }
+
+    /**
+     * Actually clear local object statistics by the given key.
+     *
+     * @param keyMsg Key to clear statistics by.
+     */
+    private void clearObjectStatisticsLocal(StatsKeyMessage keyMsg) {
+        StatsKey key = new StatsKey(keyMsg.schema(), keyMsg.obj());
+        String[] colNames = keyMsg.colNames().toArray(new String[0]);
+
         statsRepos.clearLocalPartitionsStatistics(key, colNames);
         statsRepos.clearLocalStatistics(key, colNames);
         statsRepos.clearGlobalStatistics(key, colNames);
     }
 
+    // TODO
+    /**
+     * Collect local object statistics by primary partitions of specified object.
+     *
+     * @param keyMsg Statistic key message to collect statistics by.
+     * @param partIds Set of partition ids to collect statistics by.
+     * @param cancelled Supplier to check if operation was cancelled.
+     * @return Tuple of Collected local statistics with array of successfully collected partitions.
+     * @throws IgniteCheckedException
+     */
+    private IgniteBiTuple<ObjectStatisticsImpl, int[]> collectLocalObjectStatistics(
+            StatsKeyMessage keyMsg,
+            int[] partIds,
+            Supplier<Boolean> cancelled
+    ) throws IgniteCheckedException {
+        GridH2Table tbl = schemaMgr.dataTable(keyMsg.schema(), keyMsg.obj());
+        if (tbl == null)
+            throw new IgniteCheckedException(String.format("Can't find table %s.%s", keyMsg.schema(), keyMsg.obj()));
+
+        Column[] selectedCols = filterColumns(tbl.getColumns(), keyMsg.colNames());
+
+        Collection<ObjectPartitionStatisticsImpl> partsStats = collectPartitionStatistics(tbl, partIds, selectedCols,
+                cancelled);
+        if (partsStats == null) {
+            assert cancelled.get() : "Error collecting partition level statistics.";
+
+            return null;
+        }
+
+        sendPartStatsToBU(tbl, partsStats);
+
+        StatsKey key = new StatsKey(keyMsg.schema(), keyMsg.obj());
+        statsRepos.saveLocalPartitionsStatistics(key, partsStats);
+
+        ObjectStatisticsImpl tblStats = aggregateLocalStatistics(tbl, selectedCols, partsStats);
+        statsRepos.mergeLocalStatistics(key, tblStats);
```

Comment:
Is aggregated statistics can be updated only after all partitions were processed only?

What if we got updates from primary? Will we update aggregates instantly?
If so, why we can't update aggregates after every single local partition is processed?

Reply:
"Is aggregated statistics can be updated only after all partitions were processed only?"
Yes and no. Yes, because aggregated statistics  should only contains specified in collection request partitions (to allow correctly aggregate global statistics). And no, because as local statistics it can became outdated (if some partition left current node during collection). I don't cover such scenario.
"What if we got updates from primary?"
Local level statistics contains statistics only from primary partitions. So if node will receive any partition propagation from another nodes - it won't affect local statistics at all.